### ISSUE TITLE
Faster, tighter inverse for `MixtureGaussianCDF` — Newton-in-the-loop + per-dim brackets + `einx`/`jaxtyping`

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/gauss_flows#114]
> **Type:** enhancement / performance
> **Area:** `src/gauss_flows/_src/transforms/bijections/elementwise/mixture_cdf.py`
> **Status:** proposal — ready for implementation
> **Overlap check:** searched open issues for `mixture | cdf | inverse | bisect | root | newton | spline | tail | marginal | rbig | implicit`; no overlap (#70 piecewise-linear spline is the closest neighbour but a different bijection).

---

## Summary

The current inverse path for `MixtureGaussianCDF` is **already** float32-tail-stable (PR #109, `_ndtri_exp` + log-CDF/log-SF) and **already** carries correct implicit-function-theorem gradients (PR #113, the `bisection_inverse` helper's detached-Newton trick). What it is **not** yet is *fast*: the inner loop is 50 iterations of plain bisection seeded from a global `[-100, 100]` bracket that is then expanded by doubling for 16 more iterations — a total of 66 `fori_loop` iterations to converge to ~2⁻⁵⁰ bracket width.

This issue proposes three surgical upgrades:

1. **Safeguarded Newton in the inner loop.** The mixture PDF (`_gmm_logpdf`) is already computed for the forward log-det, so a Newton step is essentially free. With a bisection fallback when Newton leaves the bracket, convergence drops from ~50 to ~5–10 iterations.
2. **Tighter initial bracket from the mixture parameters.** Replace the constants `[-100, 100]` + 16 expansion iterations with `lower = min_k(μ_k − c·σ_k)`, `upper = max_k(μ_k + c·σ_k)` (per dim). Skips the entire expansion phase.
3. **`einx` + `jaxtyping` modernisation** of the marginal block — explicit named-axis ops and array-shape annotations to match the repo's typing gate. Drops one-off `x[:, None]` broadcasts.

This is a per-coordinate marginal bijection, so the composition into `gaussianization_flow`, `iterative_rbig`, and `fit_rbig` is unchanged.

---

## Audit of the current code (with line refs)

Line refs against `src/gauss_flows/_src/transforms/bijections/elementwise/mixture_cdf.py` (354 lines, the `MixtureGaussianCDF` class) and `src/gauss_flows/_src/utils.py::bisection_inverse` (92 lines).

| Claim in earlier drafts of this proposal | Status against actual code |
|---|---|
| "Inverse is pure bisection." | **Partially true.** The inverse calls `bisection_inverse` in `utils.py`, which is 16 `fori_loop` bracket-expansion iterations + 50 `fori_loop` bisection iterations + 1 detached-Newton step at the end. Newton is *not yet used inside the loop*. |
| "Bracket may have the `min/min` bug from the `treex` ancestor." | **Not present.** The current code uses scalar bracket defaults `lower=-100.0`, `upper=100.0` (`utils.py:25–27`) and expands them by doubling. No per-coord `μ ± cσ` bracket at all — the proposed tighter bracket would be a genuine improvement, but there is no bug to fix. |
| "Confirm `lax.while_loop` / `fori_loop` and per-element bracket tracking." | **Already correct.** `bisection_inverse` uses `lax.fori_loop` (`utils.py:70, 80`), per-element brackets via `jnp.where`. Fully `jit`/`vmap` safe. |
| "No `custom_vjp`; gradients probably flow through solver iterates." | **Stale.** PR #113 added IFT-correct gradients via a *detached Newton step* trick: `x_star = stop_gradient((lo + hi) / 2)`; one Newton correction at the end attaches `dx/dy = 1/fn'(x*)` and `dx/dθ = −∂_θ fn(x*) / fn'(x*)` without `custom_vjp` (`utils.py:82–91`). Verified by `tests/test_transforms_marginal.py::test_mixture_gaussian_cdf_inverse_grad_matches_implicit` and `…_inverse_grad_through_params`. Switching to `custom_vjp` would be a wash; the existing mechanism is simpler. |
| "Verify `z` is clamped to `[eps, 1−eps]`." | **Handled differently.** `MixtureGaussianCDF` bypasses linear-CDF clamping entirely: it works in **log-space** (`_gmm_logcdf` / `_gmm_logsf`, lines 154–166) and uses a custom `_ndtri_exp` Newton solver on `log_ndtr` (lines 23–46) so the probit stays valid past ±5.5σ where naive `ndtri(exp(...))` collapses in float32. PR #109 was the fix. `MixtureLogisticCDF` *does* still use linear clamping `[1e-6, 1−1e-6]` (line 321, 336) — that's the only place a tail-stable rewrite is still wanted. |
| "Newton denominator guarded." | **Already.** `utils.py:89–90` masks `df_star != 0.0` and falls back to `x_star` when the gradient underflows. |

In short, the proposal collapses to three real items: **Newton inside the loop**, **tighter brackets**, and **`einx`/`jaxtyping` modernisation**, plus an analogous tail-stable rewrite for the *logistic* sibling.

---

## Mathematical background

For one coordinate with mixture params `{w_k, μ_k, σ_k}`:

```
Forward:   z = F(x) = Σ_k w_k Φ((x − μ_k)/σ_k)               ∈ (0, 1)
Density:   F'(x) = p(x) = Σ_k w_k φ((x − μ_k)/σ_k)/σ_k       > 0
Log-det:   log|dz/dx| = log p(x)
Inverse:   x = F⁻¹(z) solves F(x) − z = 0  (no closed form)
```

`F` is strictly increasing ⇒ the root is unique and bracketed by `[min_k(μ_k − cσ_k), max_k(μ_k + cσ_k)]`.

The existing tail-stable inverse solves `log_ndtr(y) = log F(x)` in the lower tail and `log_ndtr(−y) = log SF(x)` in the upper tail rather than `F(x) = Φ(y)` directly — keep this; do not regress to the linear path.

---

## API proposal — additive, defaults preserve current behaviour

```python
from typing import Literal
from jaxtyping import Array, Float

class MixtureGaussianCDF(AbstractBijection):
    # existing fields unchanged
    # new (with current-behaviour-preserving defaults):
    inverse_method: Literal["bisection", "newton_safe"] = "bisection"
    inverse_tol: float = 1e-6           # only the new path consults this
    inverse_max_iters: int = 50          # current path uses 50 too
    bracket_scale: float = 20.0          # only the new path consults this
```

```python
# utils.py — extended signature, all current call sites unaffected
def bisection_inverse(
    fn, y: ArrayLike,
    *,
    lower: float = -100.0, upper: float = 100.0,
    n_iter: int = 50,
    method: Literal["bisection", "newton_safe"] = "bisection",
    df: Callable | None = None,
) -> Array: ...
```

* `bisection` is the current behaviour; default unchanged.
* `newton_safe` requires `df` (`fn`'s derivative, e.g. `mixture_logpdf`) and the per-dim bracket-from-params path described below.

`MixtureLogisticCDF` gains the same kwarg and uses the analogous logistic PDF for the Newton step.

---

### (i) Tighter bracket inside `MixtureGaussianCDF.inverse_and_log_det`

```python
# replace the implicit -100/100 → expansion sequence with the explicit
# mixture-aware bracket, computed once per call:
scales_per_dim = self._scales()        # (D, K), softplus(log_scales)+5e-3
lower = jnp.min(self.means - bracket_scale * scales_per_dim, axis=-1)   # (D,)
upper = jnp.max(self.means + bracket_scale * scales_per_dim, axis=-1)   # (D,)
# (then bisect/Newton on log_ndtr(target) = log_gmm_cdf(x) per dim)
```

### (ii) Safeguarded Newton inside `bisection_inverse`

```python
def _step(_, state):
    lo, hi = state
    mid    = 0.5 * (lo + hi)
    f_mid, df_mid = jax.jvp(fn, (mid,), (jnp.ones_like(mid),))
    r      = f_mid - y
    # Newton candidate, guarded against tiny df:
    x_newt = mid - r / jnp.where(jnp.abs(df_mid) > 1e-30, df_mid, 1.0)
    in_br  = (x_newt > lo) & (x_newt < hi)
    x_next = jnp.where(in_br, x_newt, mid)
    f_next = fn(x_next)
    # bracket update uses the chosen step's sign:
    lo = jnp.where(f_next < y, x_next, lo)
    hi = jnp.where(f_next < y, hi,     x_next)
    return lo, hi
```

The detached-Newton-step IFT trick at the end (`utils.py:82–91`) is kept verbatim — it already gives the exact implicit-function gradients.

### (iii) `einx` + `jaxtyping` annotations

```python
from jaxtyping import Array, Float
import einx

def _gmm_logcdf(
    self,
    x:       Float[Array, " D"],
    means:   Float[Array, "D K"],
    scales:  Float[Array, "D K"],
    weights: Float[Array, "D K"],
) -> Float[Array, " D"]:
    log_cdfs = jstats.norm.logcdf(einx.rearrange("d, d k -> d k", x), loc=means, scale=scales)
    return einx.logsumexp("d [k]", jnp.log(weights) + log_cdfs)
```

(Apply analogously to `_gmm_logsf`, `_gmm_logpdf`.)

---

## Tests (extends what already exists)

Existing tests to keep green:

* `tests/test_transforms_marginal.py::test_mixture_gaussian_cdf_shape` / `…_finite`
* `tests/test_transforms_marginal.py::test_mixture_gaussian_cdf_inverse_grad_matches_implicit` — verifies `dx/dy = 1 / fn'(x*)` from the detached-Newton trick
* `tests/test_transforms_marginal.py::test_mixture_gaussian_cdf_inverse_grad_through_params`
* `tests/test_transforms_marginal.py::test_mixture_logistic_cdf_inverse_grad_matches_implicit`

New tests to add:

```python
def test_inverse_newton_safe_matches_bisection():
    """newton_safe and bisection converge to the same x within tol."""
    ...

def test_inverse_newton_safe_iters_lt_bisection():
    """At tol=1e-6, newton_safe needs materially fewer iters than bisection."""
    ...

def test_per_dim_bracket_tighter_than_default():
    """min_k(μ−cσ) / max_k(μ+cσ) is within an order of magnitude of
    the true root, vs ±100 for the constant bracket."""
    ...

def test_mixture_logistic_cdf_tail_stable_float32():
    """Mirror the MixtureGaussianCDF float32 tail-stability test for
    the logistic sibling once it gains the log-space path."""
    ...
```

---

## TODOs until completion

* [ ] Add `inverse_method` / `inverse_tol` / `inverse_max_iters` / `bracket_scale` fields on `MixtureGaussianCDF` and `MixtureLogisticCDF` (defaults reproduce current behaviour).
* [ ] Add per-dim `lower / upper = min_k(μ ∓ cσ)` bracket computation; route it into `bisection_inverse` via the existing `lower`/`upper` kwargs (already accepts arrays).
* [ ] Add the `newton_safe` branch to `bisection_inverse` with a `df` callable (or `jax.jvp` if `fn` is differentiable) and a bracket-aware safeguard. Keep `bisection` (default) byte-identical to today.
* [ ] Migrate `_gmm_logcdf` / `_gmm_logsf` / `_gmm_logpdf` to `einx` and add `Float[Array, "D K"]` annotations on the internal helpers.
* [ ] Add analogous log-space tail-stable path to `MixtureLogisticCDF` (use `log_sigmoid` / `log1mexp` to avoid linear clamping).
* [ ] New tests above + smoke `jit` / `vmap` / `grad` on CPU & GPU, float32 and float64. Keep existing IFT-gradient tests green.
* [ ] Benchmark iters-to-tol & wall-clock: bisection vs newton_safe across `K ∈ {2, 5, 10}`, `D ∈ {2, 64, 784}`.
* [ ] Pass all four gates: `pytest`, `ruff check .`, `ruff format --check .`, `ty check src/gauss_flows`. Update `CHANGELOG.md`.
* [ ] (Optional) Factor a shared scalar monotone-root-finder reused by `MixtureLogisticCDF` and future spline-tail inverses (see #70).

---

## Acceptance criteria

* `newton_safe` reaches `tol=1e-6` in materially fewer iterations than the current 16-expansion + 50-bisection path (~5–10 vs ~50 for the main loop; bracket expansion eliminated entirely with the per-dim bracket).
* All four pre-commit gates pass.
* Existing `transform → inverse → transform` round-trips and existing IFT gradient tests (`…_inverse_grad_matches_implicit`, `…_inverse_grad_through_params`) remain green.
* Behaviour is byte-identical to `main` when `inverse_method="bisection"` (i.e. the default), so no caller is surprised.

---

## What datasets does this make sense for?

The mixture-CDF marginal block is the core of RBIG / Gaussianization-Flows-style models, so it's strongest where those are:

* **Tabular / scientific, low-to-moderate D (≈2–100): the sweet spot.** UCI tabular density (POWER, GAS, HEPMASS, MINIBOONE, BSDS300), physical-sciences and geoscience/remote-sensing data (RBIG's traditional home), and anywhere you want **exact per-layer likelihood** and the package's **information-theoretic readouts** (`entropy`, `total_correlation`, `mutual_information`, `negentropy`). Per-coordinate multimodality is exactly what the Gaussian-mixture marginal captures, and it's where a robust/accurate inverse matters most (multimodal `F` has flat regions that punish bisection and reward Newton).
* **Images — qualified.** Raw high-D pixels (784-D MNIST, 3072-D CIFAR) are **not** the natural home: Gaussianization-style whitening needs depth growing ~linearly in dimension (Draxler et al. 2022) and lacks the convolutional inductive bias of diffusion/CNF image models. You *can* run it (the marginal+rotation stack is dimension-agnostic) but expect it to be dominated by rotation cost and uncompetitive on sample quality. It **does** make sense on **patch- or feature/latent-space** image representations (wavelet/PCA-reduced or encoder features), modest-resolution vectors, and **OOD/anomaly or information-theoretic analysis** of image features — i.e. moderate-D vectors where you want exact likelihood / information measures, not photorealistic samples. For full-res image *generation*, prefer score-based / flow-matching models and keep mixture-CDF Gaussianization for the density-estimation/analysis regime.

Bottom line: optimise this inverse for the **tabular / scientific / moderate-D** regime it is actually used in; document images as a "reduced-D / feature-space only" use case.

---

## References

* PR #109 — float32 tail-stable inversion (`_ndtri_exp` + log-CDF/log-SF).
* PR #113 — IFT-correct gradients via detached-Newton step in `bisection_inverse`.
* Meng, Song & Ermon. *Gaussianization Flows.* AISTATS 2020. arXiv:2003.01941.
* Laparra, Camps-Valls & Malo. *Iterative Gaussianization: From ICA to Random Rotations (RBIG).* IEEE TNNLS 2011.
* Chen & Gopinath. *Gaussianization.* NeurIPS 2000.
* Draxler, Schnörr & Köthe. *Whitening Convergence Rate of Coupling-based Normalizing Flows.* NeurIPS 2022; *On the Convergence Rate of Gaussianization with Random Rotations*, arXiv:2306.13520.
* Blondel et al. *Efficient and Modular Implicit Differentiation.* NeurIPS 2022 (IFT through solvers; `jaxopt`).
* Ward. *FlowJAX: Distributions and Normalizing Flows in JAX.* (base bijection interface)
* Tooling: `einx` (named-axis ops); `jaxtyping` (+ runtime checks) for array contracts.
